src/util/threadpool.rs

```diff
@@ -331,16 +341,17 @@ impl<Q, T> ThreadPool<Q, T>
             task_pool: task_pool,
             threads: threads,
             task_count: task_count,
+            sender: sender,
         }
     }
 
     pub fn execute<F>(&mut self, gid: T, job: F)
         where F: FnOnce() + Send + 'static
     {
-        let &(ref lock, ref cvar) = &*self.task_pool;
-        let mut meta = lock.lock().unwrap();
-        meta.push_task(gid, job);
+        let task = Task::new(gid, job);
+        self.sender.send(task).unwrap();
```

Comment:
It doesn't check the queue status before pushing task into it.
Is it possible that the tasks come too fast so that memory used by the channel blasts?
It's easy to limit the task pushing speed in the old way. And it's worth consider the sending/receiving speed mismatch issue.